README.md
Christof's XBMC add-on repository
=================================

Welcome to my XBMC repository!

This repository contains some awesome video add-ons for your XBMC.


plugin.video.tvbox/changelog.txt
[B]v1.0.7[/B] (January 2014) :
  - fixed broken channels
  - improved performance when loading the EPG
  - several minor bug fixes

[B]v1.0.6[/B] (July 2013) :
  - added new channels to Austria (3 channels)
        ORFSPORT+
        ATV
        SERVUSTV
  - added new channels to Germany (5 channels)
        NDR
        DAS VIERTE
        PHOENIX
        JOIZ
        ANIXE
  - added new channels to France (4 channels)
        GULLI
        GRANDLILLE
        ALSACE20
        AIR
  - added a new country: Spain (8 channels)
        KISS TV
        TELERED 21
        EUROSPORT
        CBL CLASSIC HD
        ANDALUCIA TELEVISION
        TELEMADRID SAT
        TVES
        IB3
  - added new channels to Switzerland (3 channels)
        SFINFO
        3PLUS
        TELEZUERI
  - added new channels to United Kingdom (4 channels)
        EUROSPORT
        EUROSPORT HD
        EUROSPORT2
        EUROSPORT2 HD
  - added a new country: United States of America (5 channels)
        MTV
        REDBULLTV
        5USA+1
        CBS REALITY
        CBS DRAMA

[B]v1.0.5[/B] (July 2013) :
  - fixed a bug with the EPG support of the channels from Germany

[B]v1.0.4[/B] (May 2013) :
  - improved performance when loading the countries
  - added the possibility to disable/enable EPG support for faster loading
  - fixed a bug with the playback of the channels from Germany

[B]v1.0.3[/B] (April 2013) :
  - fixed a bug with the channels from Germany
  - added new channels (7 channels)
        ZDF
        RTL NITRO
        SAT1 GOLD
        WDR
        MDR
        SWR
        N24

[B]v1.0.2[/B] (February 2013) :
  - improved EPG support and EPG layout
  - removed some non-working channels (11 channels)
        ZDF
        RTLNITRO
        MTV
        WDR
        MDR
        SWR
        N24
        SERVUSTV
        SINFO
        TELEZURI
        3PLUS
  - improved layout (channel selection by country)
  - added new channels (36 channels)
        EURONEWS GERMAN
        FRANCE24 FRENCH
        EURONEWS FRENCH
        RTL TELE LETZEBUERG
        RTP1
        RTP2
        SIC NOTICIAS
        SIC MULHER
        SIC RADICAL
        TVI FICCAO
        BENFICA TV
        PORTO CANAL
        DISNEY CHANNEL
        CANAL HOLLYWOOD
        AXN
        AXN BLACK
        AXN WHITE
        SYFY
        FOX
        FOX CRIME
        FOX MOVIES
        MTV PORTUGAL
        VH1
        DISCOVERY CHANNEL
        HISTORIA
        NATIONAL GEOGRAPHIC
        EURONEWS PORTUGUESE
        SPORT TV 1
        SPORT TV 4
        ALJAZEERA
        BBC WORLD NEWS
        BLOOMBERG
        CNN INTERNATIONAL
        CNBC EUROPE
        EURONEWS ENGLISH
        FRANCE24 ENGLISH

[B]v1.0.1[/B] (February 2013) :
  - added EPG support for the 40 channels

[B]v1.0.0[/B] (February 2013) :
  - initial release (40 channels)
        ARD
        ZDF
        SAT1
        RTL
        RTL2
        PRO7
        KABEL1
        VOX
        SIXX
        DMAX
        SUPERRTL
        RTLNITRO
        3SAT
        EUROSPORT
        SPORT1
        NICKCC
        KIKA
        ARTE
        VIVA
        MTV
        DELUXEMUSIC
        WDR
        MDR
        SWR
        N24
        NTV
        ZDFNEO
        SERVUSTV
        ORF1
        ORF2
        SF1
        SF2
        SFINFO
        TELEZURI
        3PLUS
        RTSUN
        RTSDEUX
        FRANCE2
        FRANCE3
        FRANCE5


plugin.video.tvbox/test.py
# -*- coding: utf-8 -*-

#TV Box - by Christof Torres 2012 - 2013.

import urllib,urllib2,re,datetime,os.path

from xml.dom import minidom

user_agent = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3'

xml = 'resources/data/channels_de.xml'

number = 0
xmldoc = minidom.parse(xml)
channel_list = xmldoc.getElementsByTagName('channel')
for channel in channel_list:
        number += 1
        name = channel.attributes['name'].value.encode("utf-8")
        url = channel.attributes['url'].value.encode("utf-8")
        try:
                if (addon.getSetting("epgsupport") == "true"):
                        epg = get_epg(channel.attributes['epg'].value.encode("utf-8"))
                else:
                        epg = ''
        except:
                epg = ''
        try:
                if (url.find('www.tv-kino.net') != -1):
                        req = urllib2.Request(url)
                        req.add_header('User-Agent', user_agent)
                        response = urllib2.urlopen(req)
                        link = response.read()
                        response.close()
                        flashvars = re.compile('file=(.+?)&amp;.+?streamer=(.+?)&amp;').findall(link)
                        for playpath, rtmp in flashvars:
                                rtmp = rtmp+' swfUrl=http://stream.tv-kino.net/player.swf playpath='+playpath+' pageurl='+url+' live=true swfvfy=true'
                elif (url.find('megatv.to') != -1):
                        req = urllib2.Request(url)
                        req.add_header('User-Agent', user_agent)
                        response = urllib2.urlopen(req)
                        link = response.read()
                        response.close()
                        #rtmp = re.compile('<video width="80%" src="(.+?)" controls="controls"></video>').findall(link)[0].replace("&amp;", "&")
                        #rtmp = re.compile('\'file\': \'(.+?)\'').findall(link)[0].replace("&amp;", "&")
                        rtmp = 'http://sv1.megatv.to/index.m3u8?c=pro7&auth=5e13d234495fef040feed7a17358c97a'
                else:
                        rtmp = url
                        thumbnail = '/resources/media/'+name.lower()+'.jpg'
                print number
                print name
                print epg
                print rtmp
                print thumbnail
                print len(channel_list)
        except:
                print "Unexpected Error: "+url



plugin.video.kinokiste/test.py
import urllib,urllib2,re

#KinoKiste - by Christof Torres 2011 - 2013.


user_agent = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3'
                        
url = 'http://www.kkiste.to/aktuelle-kinofilme/?page=1'

main_url = url
req = urllib2.Request(url)
req.add_header('User-Agent', user_agent)
response = urllib2.urlopen(req)
link = response.read()
response.close()
pages = re.compile('<li><a href="\?page=.+?">(.+?)</a></li>').findall(link)
match = re.compile('<a href="(.+?)" title="(.+?)" class="title">.+?</a>').findall(link)
for url,name in match:
                try:
                        url = 'http://www.kkiste.to'+url
                        req3 = urllib2.Request(url)
                        req3.add_header('User-Agent', user_agent)
                        response3 = urllib2.urlopen(req3)
                        link3 = response3.read()
                        response3.close()
                        name = name[6:len(name)]
                        name = name[0:name.find('Stream ansehen')-1]
                        thumbnail = re.compile('<img src="(.+?)" width="145" height="215" alt=".+?" />').findall(link3)[0]
                        director = re.compile('<p><span>von:</span> <a href=".+?">(.+?)</a>').findall(link3)[0]
                        rating = re.compile('IMDB Rating:</span> <a href=".+?" target="_blank" title=".+?">(.+?)</a>').findall(link3)[0]
                        year = re.compile('Jahr:</span> (.+?)</p>').findall(link3)[0] 
                        plot = re.compile('<p id="kk-plot">(.+?)<span class="kk-ds">').findall(link3)[0].replace("<br>","\n").replace("<br />","\n").replace("&nbsp;"," ")
                        actors = re.compile('<li><a href="/darsteller/.+?/" title=".+?">(.+?)</a></li>').findall(link3)
                        cast = ''
                        for actor in actors:
                                cast = cast+actor+'\n'
                        if (main_url[20:len(main_url)] == '/serien/'):
                                addDir(name,url,8,thumbnail)
                        else:
                                print name
                except:
                        print 'Unexpected error: '+url
last_page = pages[len(pages)-1]
current_page = main_url[46:len(main_url)]
print current_page
print last_page
                        


plugin.video.tvbox/default.py
# -*- coding: utf-8 -*-

#TV Box - by Christof Torres 2012 - 2014.

import urllib,urllib2,re,datetime,os.path,xbmcplugin,xbmcgui,xbmcaddon

from xml.dom import minidom

addon = xbmcaddon.Addon(id='plugin.video.tvbox')

user_agent = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3'

root = 'https://raw.github.com/ChristofTorres/christofs-xbmc-addon-repository/master/plugin.video.tvbox/resources/data/countries.xml'

def COUNTRIES():
        xmlfile = urllib2.urlopen(root)
        error_server = 0
        if (os.path.exists(addon.getAddonInfo('path')+'/resources/data/'+root.split('/')[-1])):
                xmldocserver = minidom.parse(xmlfile)
        else:
                output = open(addon.getAddonInfo('path')+'/resources/data/'+root.split('/')[-1],'wb')
                output.write(xmlfile.read())
                output.close()
                xmldocserver = minidom.parse(addon.getAddonInfo('path')+'/resources/data/'+root.split('/')[-1])
                error_server = -1
        xmldoclocal = minidom.parse(addon.getAddonInfo('path')+'/resources/data/'+root.split('/')[-1])
        country_list_local = xmldoclocal.getElementsByTagName('country')
        index = -1; error_local = 0
        for country_server in xmldocserver.getElementsByTagName('country'):
                index += 1
                country_found = False
                for country_local in country_list_local:
                        if (country_server.attributes['name'].value == country_local.attributes['name'].value):
                                country_found = True
                if (index < len(country_list_local)):
                        timestamp_server = country_server.attributes['timestamp'].value.encode("utf-8")
                        timestamp_local = country_list_local[index].attributes['timestamp'].value.encode("utf-8")
                xml = country_server.attributes['xml'].value.encode("utf-8")
                if (not country_found or not os.path.exists(addon.getAddonInfo('path')+'/resources/data/'+xml.split('/')[-1]) or timestamp_server != timestamp_local):
                        xmlfile = urllib2.urlopen(xml)
                        output = open(addon.getAddonInfo('path')+'/resources/data/'+xml.split('/')[-1],'wb')
                        output.write(xmlfile.read())
                        output.close()
                        error_local = -1
                xml = addon.getAddonInfo('path')+'/resources/data/'+xml.split('/')[-1]
                name = '[B]'+country_server.attributes['name'].value.encode("utf-8")+'[/B] ('+str(len(minidom.parse(xml).getElementsByTagName('channel')))+')'
                thumbnail = addon.getAddonInfo('path')+'/resources/media/'+country_server.attributes['name'].value.encode("utf-8").lower()+'.jpg'
                addDir(name, xml, 1, thumbnail, 0)
        if (error_server != -1 and error_local == -1):
                xmlfile = urllib2.urlopen(root)
                output = open(addon.getAddonInfo('path')+'/resources/data/'+root.split('/')[-1],'wb')
                output.write(xmlfile.read())
                output.close()
                

def TVCHANNELS(xml):
        number = 0
        xmldoc = minidom.parse(xml)
        channel_list = xmldoc.getElementsByTagName('channel')
        for channel in channel_list:
                number += 1
                name = channel.attributes['name'].value.encode("utf-8")
                url = channel.attributes['url'].value.encode("utf-8")
                try:
                        if (addon.getSetting("epgsupport") == "true"):
                                if (url.find('megatv.to') == -1):
                                       epg = get_epg(channel.attributes['epg'].value.encode("utf-8"))
                        else:
                                epg = ''
                except:
                        epg = ''
                try:
                        if (url.find('www.live-stream.tv') != -1):
                                req = urllib2.Request(url)
                                req.add_header('User-Agent', user_agent)
                                response = urllib2.urlopen(req)
                                link = response.read()
                                response.close()
                                flashvars = re.compile('file=(.+?)&amp;.+?streamer=(.+?)&amp;').findall(link)
                                for playpath, rtmp in flashvars:
                                        rtmp = rtmp+' swfUrl=http://stream.live-stream.tv/player.swf playpath='+playpath+' pageurl='+url+' live=true swfvfy=true'
                        elif (url.find('megatv.to') != -1):
                                req = urllib2.Request(url)
                                req.add_header('User-Agent', user_agent)
                                response = urllib2.urlopen(req)
                                link = response.read()
                                response.close()
                                rtmp = re.compile('\'file\': \'(.+?)\'').findall(link)[0]
                                try:
                                        epg_title = re.compile('<center><font color=\'black\' size=\'4\' face=\'Jockey One\'>(.+?)<br>').findall(link)
                                        epg_time = re.compile('von (.+?) bis (.+?)</font><br>').findall(link)
                                        if (len(epg_title) > 0 and len(epg_time) > 0):
                                                epg = epg_title[0]+" "+epg_time[0][0]+" - "+epg_time[0][1]
                                        else:
                                                epg = get_epg(channel.attributes['epg'].value.encode("utf-8"))
                                except:
                                        epg = ''
                        else:
                                rtmp = url
                        thumbnail = addon.getAddonInfo('path')+'/resources/media/'+name.lower()+'.jpg'
                        addLink(number, name, epg, rtmp, thumbnail, len(channel_list))
                except:
                        print "Unexpected Error: "+url


def get_epg(url):
        req = urllib2.Request(url)
        req.add_header('User-Agent', user_agent)
        response = urllib2.urlopen(req)
        link = response.read()
        response.close()
        if (url.find('http://i.teleboy.ch') <> -1):
                now = link.split('<div id="showbox__0" class="showclass">')
                epg = re.compile('<p class="show_title"><a href="/programm/detail.php\?const_id=.+?">(.+?)</a>(.+?)<br /></p>').findall(now[1])
                title = epg[0][0].decode("iso-8859-1")
                title = title.encode("utf-8") 
                time = epg[0][1].decode("iso-8859-1")
                time = time.encode("utf-8")
        elif (url.find('tele.rtl.lu') <> -1):
                now = re.compile('<TR><TD class="highlight">(.+?)</TD><TD>&nbsp;<B><A HREF=".+?>(.+?)</A>.+?</B></TD></TR>').findall(link)
                epg = re.compile('<TR><TD class="highlight">(.+?)</TD><TD>&nbsp;.*?<A HREF=".*?>(.+?)</A>.+?</TD></TR>').findall(link)
                index = -1
                for epg_time, epg_title in epg:
                        index += 1
                        if (epg_time == now[0][0] and epg_title == now[0][1]):
                                index += 1
                                time = now[0][0]+' - '+epg[index][0]
                                title = now[0][1]
                                break
        elif (url.find('portalnacional.com.pt') <> -1):
                epg = re.compile('<div class="data">(.+?)</div><div class="titulo"><a href="(.+?)" title=".+?">(.+?)</a></div>').findall(link)
                now = datetime.datetime.utcnow()
                now_dt = datetime.datetime(now.year, now.month, now.day, now.hour, now.minute, 0, 0)
                for index in range(0, len(epg)-1):
                        epg[index] = (epg[index][0],epg[index+1][0],epg[index][2])
                for epg_start_time, epg_end_time, epg_title in epg:
                        epg_start_dt = datetime.datetime(now.year, now.month, now.day, int(epg_start_time.split(':')[0]), int(epg_start_time.split(':')[1]), 0, 0)
                        epg_end_dt = datetime.datetime(now.year, now.month, now.day, int(epg_end_time.split(':')[0]), int(epg_end_time.split(':')[1]), 0, 0)
                        if (epg_start_dt.time() <= now_dt.time() and epg_end_dt.time() >= now_dt.time()):
                                time = epg_start_time+' - '+epg_end_time
                                title = epg_title
                                break
        else:
                print 'Unknown EPG: '+url       
        return title+' '+time


def get_params():
        param=[]
        paramstring=sys.argv[2]
        if len(paramstring)>=2:
                params=sys.argv[2]
                cleanedparams=params.replace('?','')
                if (params[len(params)-1]=='/'):
                        params=params[0:len(params)-2]
                pairsofparams=cleanedparams.split('&')
                param={}
                for i in range(len(pairsofparams)):
                        splitparams={}
                        splitparams=pairsofparams[i].split('=')
                        if (len(splitparams))==2:
                                param[splitparams[0]]=splitparams[1]
                                
        return param


def addLink(number,name,epg,url,iconimage,totalchannels):
        ok=True
        if (number < 10):
                name = '[B]'+'  '+str(number)+' '+name+'[/B]'+'      '+epg
        else:
                name = '[B]'+str(number)+' '+name+'[/B]'+'      '+epg
        liz=xbmcgui.ListItem(name, iconImage="DefaultVideo.png", thumbnailImage=iconimage)
        liz.setInfo( type="Video", infoLabels={ "Title": epg } )
        ok=xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]),url=url,listitem=liz,isFolder=False,totalItems=totalchannels)
        return ok


def addDir(name,url,mode,iconimage,totalcountries):
        u=sys.argv[0]+"?url="+urllib.quote_plus(url)+"&mode="+str(mode)+"&name="+urllib.quote_plus(name)
        ok=True
        liz=xbmcgui.ListItem(name, iconImage="DefaultFolder.png", thumbnailImage=iconimage)
        liz.setInfo( type="Video", infoLabels={ "Title": name } )
        ok=xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]),url=u,listitem=liz,isFolder=True,totalItems=totalcountries)
        return ok
        
              
params=get_params()
url=None
name=None
mode=None

try:
        url=urllib.unquote_plus(params["url"])
except:
        pass
try:
        name=urllib.unquote_plus(params["name"])
except:
        pass
try:
        mode=int(params["mode"])
except:
        pass

print "Mode: "+str(mode)
print "URL : "+str(url)
print "Name: "+str(name)

if mode==None or url==None or len(url)<1:
        print "COUNTRIES()"
        COUNTRIES()

elif mode==1:
        print "TVCHANNELS("+url+")"
        TVCHANNELS(url)

xbmcplugin.endOfDirectory(int(sys.argv[1]))


plugin.video.kinokiste/episode.py
import urllib,urllib2,re

#KinoKiste - by Christof Torres 2011 - 2013.


user_agent = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3'
                        
url = 'http://kkiste.to/how-i-met-your-mother-stream.html'

main_url = url
url = url+'?season=1'
req = urllib2.Request(url)
req.add_header('User-Agent', user_agent)
response = urllib2.urlopen(req)
link = response.read()
response.close()
print link
match = re.compile('value="(.+?)">Episode .+?</option>').findall(link)
print match

serie = main_url[25:main_url.find("-stream")]
serie = serie.lower()
serie = serie.replace("-", "+")

url2='http://www.fernsehserien.de/index.php?suche='+serie

req2 = urllib2.Request(url2)
req2.add_header('User-Agent', 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3')
response2 = urllib2.urlopen(req2)
link2=response2.read()
response2.close()

#link2 = link2[link2.find('staffel='+name[8:len(name)]+'">'+name+'</a>'):len(link2)]
#link2 = link2[0:link2.find("<a href=")]
#link2 = link2.replace("<br>", "<br><br>")

match2=re.compile('<br>(.+?)<br>').findall(link2)

tab_episode = [["" for col in range(2)] for row in range(int(match[0]))]

for number in match:
        index = int(number)-1
        if (len(match2) > 1):
                title = match2[index]
                title = title[title.find(" ")+1:len(title)]
                name = 'Episode '+number+' - '+title
        else:
                name = 'Episode '+number

        tab_episode[index][0] = name
        tab_episode[index][1] = url+'&episode='+number

for index in range(int(match[0])):
        try:
                print tab_episode[index][1]
                print tab_episode[index][0]
        except:
                print 'EPISODE '+str(index+1)+' IS MISSING...!'


plugin.video.kinokiste/ecostream.py
import urllib,urllib2,re

#KinoKiste - by Christof Torres 2011 - 2013.


user_agent = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3'
                        
url = 'http://kkiste.to/harry-potter-und-die-kammer-des-schreckens-stream.html'

req = urllib2.Request(url)
req.add_header('User-Agent', user_agent)
response = urllib2.urlopen(req)
link = response.read()
response.close()
parts = re.compile('<li class=".+?"><a href="(.+?)" target="_blank">Ecostream <small>(.+?)</small></a></li>').findall(link)
for part, part_number in parts:
        print part
        formdictionary = { 'ss' : '1', 'sss' : '1' }
        params = urllib.urlencode(formdictionary)
        req2 = urllib2.Request(part, params)
        req2.add_header('User-Agent', user_agent)
        response2 = urllib2.urlopen(req2)
        link2 = response2.read()
        response2.close()
        match2 = re.compile("var t=setTimeout\(\"lc\('([^']+)','([^']+)','([^']+)','([^']+)'\)\",.+?\);").findall(link2)
        for a,b,t,key in match2:
                url = 'http://www.ecostream.tv/lo/mq.php?s='+a+'&k='+b+'&t='+t+'&key='+key
                formdictionary2 = { 's': a, 'k' : b, 't' : t, 'key' : key }
                params2 = urllib.urlencode(formdictionary2)
                req3 = urllib2.Request(url, params2)
                req3.add_header('Referer', 'http://www.ecostream.tv')
                req3.add_header('X-Requested-With', 'XMLHttpRequest')
                response3 = urllib2.urlopen(req3)
                link3 = response3.read()
                response3.close()
                print link3
                stream_url = re.compile('<param name="flashvars" value="file=(.*?)&').findall(link3)[0]
                stream_url = 'http://www.ecostream.tv'+stream_url
                print stream_url
                print 'Ecostream '+part_number


plugin.video.kinokiste/test copy.py
import urllib,urllib2,re

#KinoKiste - by Christof Torres 2011 - 2013.


user_agent = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.3) Gecko/2008092417 Firefox/3.0.3'
                        
url = 'http://www.kkiste.to/aktuelle-kinofilme/'

main_url = url
req = urllib2.Request(main_url)
req.add_header('User-Agent', user_agent)
response = urllib2.urlopen(req)
link = response.read()
response.close()
menu = url[20:len(url)]
pages = re.compile('<li><a href="\?page=.+?">(.+?)</a></li>').findall(link)
if (len(pages) != 0):
        pages = int(pages[len(pages)-1])
else:
        pages = 1
for i in range(1, (2)):
        page_url = main_url+'?page='+str(i)
        req2 = urllib2.Request(page_url)
        req2.add_header('User-Agent', user_agent)
        response2 = urllib2.urlopen(req2)
        link2 = response2.read()
        response2.close()
        match = re.compile('<a href="(.+?)" title="(.+?)" class="title">.+?</a>').findall(link2)
        for url,name in match:
                try:
                        url = 'http://www.kkiste.to'+url
                        req3 = urllib2.Request(url)
                        req3.add_header('User-Agent', user_agent)
                        response3 = urllib2.urlopen(req3)
                        link3 = response3.read()
                        response3.close()
                        name = name[6:len(name)]
                        name = name[0:name.find('Stream ansehen')-1]
                        thumbnail = re.compile('<img src="(.+?)" width="145" height="215" alt=".+?" />').findall(link3)[0]
                        director = re.compile('<p><span>von:</span> <a href=".+?">(.+?)</a>').findall(link3)[0]
                        rating = re.compile('IMDB Rating:</span> <a href=".+?" target="_blank" title=".+?">(.+?)</a>').findall(link3)[0]
                        year = re.compile('Jahr:</span> (.+?)</p>').findall(link3)[0] 
                        plot = re.compile('<p id="kk-plot">(.+?)<span class="kk-ds">').findall(link3)[0].replace("<br>","\n").replace("<br />","\n").replace("&nbsp;"," ")
                        actors = re.compile('<li><a href="/darsteller/.+?/" title=".+?">(.+?)</a></li>').findall(link3)
                        cast = ''
                        for actor in actors:
                                cast = cast+actor+'\n'
                        if (menu == '/serien/'):
                                addDir(name,url,8,thumbnail)
                        else:
                                print name
                except:
                        print 'Unexpected error: '+url
                        
